Aim separation in difficulty calculation (base)

Givikap120 opened the pull request:
This is a base version of separation that mainly focuses on preserving correct balancing.
Intended to be expanded with various bonuses that will improve quality of evaluators.

Givikap120 committed:
Introduce Flow Aim eval

Givikap120 committed:
Make changes to other evals to compensate for flow eval

Givikap120 committed:
Add angle signed

Givikap120 committed:
Add new skills

Givikap120 committed:
Add new rating calculators

Givikap120 committed:
Change summation

Givikap120 committed:
Add new difficulty attributes

Givikap120 committed:
Add aim versatility bonus

Givikap120 committed:
Apply antirake to flow aim

Givikap120 committed:
Multipliers rebalance

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -36,22 +37,23 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
 
             // Calculate the velocity to the current hitobject, which starts with a base distance / time assuming the last object is a hitcircle.
-            double currVelocity = osuCurrObj.LazyJumpDistance / osuCurrObj.AdjustedDeltaTime;
+            double currDistance = osuCurrObj.LazyJumpDistance + calculateSnappingDifficulty(osuCurrObj.LazyJumpDistance, osuCurrObj, osuLastObj);
```
This is arguably semantics but for clarity and sanity's sake, could we change this so that rather than adding onto `currDistance` etc., we have a base difficulty variable that is equal to the inherent snapping distance and then add everything else on instead? To me it would be way clearer that the intent is to deal with snapping having an inherent difficulty

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Skills/TotalAim.cs:
```diff
@@ -0,0 +1,26 @@
+
+namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+{
+    public class TotalAim : Aim
```
nitpick but maybe naming this `CombinedAim` is a little clearer

Givikap120 committed:
Add reverse lerp two directional

Givikap120 committed:
Add difficulty attribute ids

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:
```diff
@@ -139,12 +145,41 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
+                );
+
+            return difficulty * multiplier;
+        }
```
Why do we need this?

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/OsuRatingCalculator.cs:
```diff
@@ -29,20 +35,93 @@ public OsuRatingCalculator(Mod[] mods, int totalHits, double approachRate, doubl
+
+            // To ensure that result would not be bigger than normal aim difficulty rating
+            if (mods.Any(m => m is OsuModTouchDevice))
+                flowAimRating = Math.Pow(flowAimRating, 0.83);
```
does this still yield a TD nerf for flow aim in the end?

tsunyoku reviewed (commented):
Some comments to get us started.

Givikap120 committed:
Remove unnecessary magnetized penalties

Givikap120 committed:
use correct function for difficulty value

Givikap120 committed:
Update OsuRatingCalculator.cs

Givikap120 commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -80,32 +82,33 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
 
+                // Rescale wide angle bonus to reward lower spacing more
+                double velocityThreshold = diameter * 2.3 / osuCurrObj.AdjustedDeltaTime;
+                double wideVelocityBase = Math.Min(angleBonus, velocityThreshold + 0.4 * (angleBonus - velocityThreshold));
```
Here the scaling for aim was changed from d/t^2.2 (roughly) to d/t^2 thanks to no distance limit in speed
Because of this - all high spaced patterns were buffed. But because wide angle bonus as very big spacing - by far the most buffed patterns were squares and c-type type stuff.
They were buffed significantly regardless of BPM (so it's now because of snapping difficulty thing), this rescale brings it back to be similar how it looked in master

Givikap120 committed:
Fix the versatility formula being applied incorrectly

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs:
```diff
@@ -125,12 +128,12 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
+            // We want to use the average velocity over the whole object when awarding differences, not the individual jump and slider path velocities.
+            prevVelocity = (osuLastObj.LazyJumpDistance + osuLastLastObj.TravelDistance) / osuLastObj.AdjustedDeltaTime;
+            currVelocity = (osuCurrObj.LazyJumpDistance + osuLastObj.TravelDistance) / osuCurrObj.AdjustedDeltaTime;
+
```
what is referencing these that wasn't before? are you implying this is a pre-existing issue? I don't understand

Givikap120 replied:
Zero velocity can be turned into non-zero thanks to "snapping difficulty"
so when raw velocity is recalculated for velocity change bonus - it can drop it back to zero again

Givikap120 committed:
Fix stated in review stuff #1

Givikap120 committed:
Return missed semicolon

Givikap120 committed:
Add angle change bonus

Givikap120 committed:
update flow multiplier

Givikap120 committed:
Balancing

Save me nerf
Crystalia+Anone nerf

Givikap120 committed:
code quality improvements

Givikap120 committed:
change summation

Givikap120 committed:
Increase rhythm pp

Givikap120 committed:
Remove unnecessary <diameter check

Givikap120 committed:
Snap angle change bonus balancing

Slight change for crystalia to be neutral so there would be no problems in future versions

Givikap120 committed:
Add anti-cheese on wiggle bonus

Givikap120 committed:
balancing

Givikap120 committed:
Fixed bug where speedflow bonus rewards zero spacing

Givikap120 committed:
Refactor snap and flow weighting

Givikap120 committed:
Replace versatility bonus with straightforward summation

Givikap120 committed:
Nerf cheesable angle change scenarios

Givikap120 committed:
Add multipliers for bonuses

Givikap120 committed:
Fix name violation

Givikap120 committed:
Balancing

Givikap120 committed:
Fix wiggle bonus to be stable

Givikap120 committed:
Rebalance flow

Givikap120 committed:
Unnerf flow aim on touch device

Givikap120 committed:
Fix typo

Givikap120 committed:
Change anti-rake for flow aim

Givikap120 committed:
Make the doubles nerf smarter in snap aim eval

Givikap120 committed:
Update comments

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Change bonuses to be multiplier-based

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Merge

Givikap120 committed:
Refactor slider bonus

Givikap120 committed:
Adjust acute bonus

Givikap120 committed:
Slight balancing

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
rework how acute angle multipliers work (again)

Givikap120 committed:
increase rhythm multi

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Update FlowAimEvaluator.cs

Givikap120 committed:
nerf acute slightly

Givikap120 committed:
Refactor

Givikap120 committed:
Rework bonuses to be based on velocity again

Givikap120 committed:
Simplify velocity calculation

Givikap120 committed:
Update comment

Givikap120 committed:
balancing

Givikap120 committed:
Use single exponent

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
revert length bonus

Givikap120 committed:
slight speed buff

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Move flow multiplier

Givikap120 committed:
Nerf speedflow slightly

Givikap120 committed:
Buff speed more

Givikap120 committed:
More balancing

Givikap120 committed:
Surely last balancing

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Use correct distance

Givikap120 committed:
Remove extra speedflow buff

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
nerf flow aim

Givikap120 committed:
Revert rhythm eval change

Givikap120 committed:
Make length bonus match confirmed

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
balancing

Givikap120 committed:
Nerf speedflow slightly

Givikap120 committed:
Add snap/flow perf attributes

Givikap120 committed:
Merge branch 'pp-dev' into aim-separation-simplified

Givikap120 committed:
Temporary remove speedflow bonus

Givikap120 committed:
Balance adjust

Givikap120 committed:
Use power in wide angle bonus instead of threshold

Givikap120 committed:
More balance

Givikap120 committed:
move distance exponent into it's own constant
